dtfs-central-api/src/helpers/fee-record-payment-entity-group.helper.ts

```diff
@@ -0,0 +1,143 @@
+  if (group.feeRecords.length === 0) {
+    throw new Error('Fee record payment entity group cannot have an empty fee records array');
+  }
+
+  const allStatusesInGroup = group.feeRecords.reduce((statuses, { status }) => statuses.add(status), new Set<FeeRecordStatus>());
+
+  if (allStatusesInGroup.size === 1) {
+    const [status] = allStatusesInGroup.values();
+    return status;
+  }
+
+  if (allStatusesInGroup.size === 2 && allStatusesInGroup.has(FEE_RECORD_STATUS.READY_TO_KEY) && allStatusesInGroup.has(FEE_RECORD_STATUS.RECONCILED)) {
+    return FEE_RECORD_STATUS.READY_TO_KEY;
+  }
+
+  const formattedGroupStatuses = Array.from(allStatusesInGroup.values()).join(', ');
+  throw new Error(`Fee record payment entity group has an invalid set of statuses: [${formattedGroupStatuses}]`);
+};
+
+/**
+ * Gets the reconciliation data for a fee record payment entity group
+ * @param group - The fee record payment entity group
+ * @returns The group reconciliation data
+ * @throws {Error} If the groups fee records array is empty
+ * @throws {Error} If any of the fee records at the `RECONCILED` status have a null `dateReconciled` property
+ */
+export const getFeeRecordPaymentEntityGroupReconciliationData = async (
+  group: FeeRecordPaymentEntityGroup,
+): Promise<{
+  reconciledByUser?: FeeRecordReconciledByUser;
+  dateReconciled?: Date;
+}> => {
+  if (group.feeRecords.length === 0) {
+    throw new Error('Fee record payment entity group cannot have an empty fee records array');
+  }
+
+  const reconciledFeeRecords = group.feeRecords.filter(({ status }) => status === FEE_RECORD_STATUS.RECONCILED);
+
+  if (reconciledFeeRecords.length === 0) {
+    return {};
+  }
+
+  if (reconciledFeeRecords.some(({ dateReconciled }) => dateReconciled === null)) {
+    throw new Error(`Fee records at the '${FEE_RECORD_STATUS.RECONCILED}' status cannot have a null 'dateReconciled' property`);
+  }
+
+  const feeRecordsSortedByDateReconciledDescending = orderBy(reconciledFeeRecords, [(feeRecord) => feeRecord.dateReconciled!.getTime()], ['desc']);
+  const mostRecentlyReconciledFeeRecord = feeRecordsSortedByDateReconciledDescending.at(0);
+  if (!mostRecentlyReconciledFeeRecord) {
+    // Based on all the above checks, this error should never be reached, hence not being documented
+    throw new Error('Something went wrong when sorting the fee records to get the most recently reconciled fee record');
+  }
+
+  const { dateReconciled, reconciledByUserId } = mostRecentlyReconciledFeeRecord;
+
+  if (!reconciledByUserId) {
+    return { dateReconciled: dateReconciled! };
+  }
+
+  const reconciledByTfmUser = await TfmUsersRepo.findOneUserById(reconciledByUserId);
+
+  if (!reconciledByTfmUser) {
+    return { dateReconciled: dateReconciled! };
+  }
+
+  const { firstName, lastName } = reconciledByTfmUser;
+
+  return {
+    dateReconciled: dateReconciled!,
+    reconciledByUser: { firstName, lastName },
+  };
+};
```

Comment:
I'd prefer to leave this as is for now as it matches our existing pattern for helpers and apply this change down the line if needed in a separate PR

Reply:
The other helpers in central API are a lot simpler, they all have 1/2 functions in each file - i'd agree if these new helpers were smaller additions to an existing file, but this is a new helper file with 4x helper functions and the unit test is nearly 400 lines long 😢 I'm thinking could be something like this, so we have smaller helpers, easier to read/digest and test?

```
helpers/fee-record-payment-entity-group/get-payment-id-key
helpers/fee-record-payment-entity-group/get-groups
```

Reply:
The procedure you've suggested highlights why we don't want to export functions just for the sake of unit tests - structuring the files in this way implies that `getPaymentIdKey` is something that should be used outside the context of generating groups, which it should not. Again, because the scope of this PR continuing to grow, I would suggest that this comes under tech debt and should be left for future work

Reply:
Every function, especially helpers, should have a unit test.

This feels like it's introducing tech debt, by having 1x huge single helper with multiple helpers in.

My suggestion for creating new directories for these helpers was just an example of how it _could_ be split up.

Reply:
I have split these into multiple test helpers and moved the `getPaymentIdKey` function to the scope of the function it is used in so that new users don't try and use it